### PR TITLE
[NestedTensor] Add support for reducing along multiple non-ragged dimensions for the mean operator in NestedTensor

### PR DESCRIPTION
**Pull request by jananisriram**
Summary: Add support for reducing across multiple non-ragged dimensions using the `mean` operator in `NestedTensor`. For example, for a nested tensor of shape `(B, *, M, N)`, this diff enables reducing across dimensions `(M, N)` to an output of shape `(B, *)`.

Test Plan:
Verify that existing and modified unit tests pass via the following commands:

```
buck2 run mode/{opt,inplace} //caffe2/test:nested -- --regex test_mean
```

```
buck2 run mode/{opt,inplace} //caffe2/test:nested -- --regex test_jagged_op
```

Differential Revision: D60680005

### CODE REVIEW

File: torch/testing/_internal/opinfo/definitions/nested.py [hunk 1]
```diff
@@ -230,10 +289,12 @@ def sample_inputs_special_polygamma_n(n):
 # to specify if they cannot be auto-generated for some reason. Try to keep these sorted
 # in alphabetical order!
 njt_sample_inputs = {
+    "native_layer_norm": sample_inputs_layer_norm,
```
Comment (cpuhrsch) [inline torch/testing/_internal/opinfo/definitions/nested.py:292]: so this PR also adds support for layer norm (as a composite) and softmax?

File: torch/nested/_internal/ops.py [hunk 2]
```diff
@@ -1438,21 +1428,45 @@ def mean_dim(func, *args, **kwargs):
...
+        if (
+            reduce_on_batch
+        ):  # invalid reduction cases: (batch), (batch, non-batch), etc.
+            raise RuntimeError(
+                "mean(): not supported along the batch dimension but not the ragged dimension for NestedTensor"
+            )
+        # reduction cases: (non-batch), (non-batch, non-batch), etc.
+        return NestedTensor(
+            func(inp._values, **new_kwargs), **extract_kwargs(inp)
+        )  # apply sum directly on values
```
Comment (soulitzer) [inline torch/nested/_internal/ops.py:1469]:
```suggestion
        )  # apply mean directly on values
```

File: torch/testing/_internal/opinfo/definitions/nested.py [hunk 3]
```diff
@@ -180,6 +180,23 @@ def sample_inputs_njt_reduction(
...
+            for keepdim in [False, True]:
+                yield SampleInput(
+                    njt,
+                    kwargs={
+                        **op_kwargs,
+                        "dim": (dim, dim + 1, dim + 2),
+                        "keepdim": keepdim,
+                    },
+                )
+
```
Comment (soulitzer) [inline torch/testing/_internal/opinfo/definitions/nested.py:199]: Do we actually have any reductions that support `keepdim=False` right now? Maybe better to just update this to only run with keepdim=True so that we get the coverage and add a todo to reenable keepdim when we start supporting that case.

File: test/test_nestedtensor.py [hunk 4]
```diff
@@ -6986,8 +6998,10 @@ def test_layout_conversion_backward(self, device):
     "var",
     "var.unbiased",
     # === BEGIN UNSUPPORTED SECTION ===
-    # RuntimeError: mean(): not supported for NestedTensor on dim=1
+    # RuntimeError: mean(): not supported when keepdim=False for NestedTensor
     "mean",
+    # NotImplementedError: aten.sum.default
+    "sum",
```
Comment (soulitzer) [inline test/test_nestedtensor.py:7004]: Any how this error changed to NotImplementedError?

File: torch/nested/_internal/ops.py [hunk 5]
```diff
@@ -661,8 +661,8 @@ def _softmax_default(func, *args, **kwargs):
     if reduce_on_ragged:
         padded_softmax_values = torch.nn.functional.softmax(
             torch.ops.aten._jagged_to_padded_dense_forward(
-                inp._values.flatten(
-                    start_dim=inp._ragged_idx
+                inp._values.reshape(
+                    inp._values.shape[0], -1
```
Comment (soulitzer) [inline torch/nested/_internal/ops.py:665]: Is this change intended to support the 2D NJT case right?

File: test/test_nestedtensor.py [hunk 6]
```diff
@@ -7014,12 +7028,12 @@ def test_layout_conversion_backward(self, device):
...
     "jiterator_binary_return_by_ref",
     "jiterator_unary",
-    # Bug found: sum() with keepdim=True returns invalid shape
-    "sum",
     # RuntimeError: prod(): keepdim=True must be set for NestedTensor
     "prod",
     # RuntimeError: "jagged_to_padded_dense" not implemented for 'Bool'
     "nanmean",
+    # AssertionError: assert not isinstance(out_ref_component, (list, tuple))
+    "native_layer_norm",
```
Comment (soulitzer) [inline test/test_nestedtensor.py:7036]: Are there any particular inputs that is problematic? 
(if that is the case, maybe worth opening a new issue with some more details, and linking back here)

File: torch/testing/_internal/opinfo/definitions/nested.py (hunk 1 above)
Comment (soulitzer) [reply torch/testing/_internal/opinfo/definitions/nested.py:292]: Softmax should be supported already, though this PR updates the logic and adds an OpInfo test.
layer_norm should be already supported as well, and this PR just adds an OpInfo test.

**Review by soulitzer — COMMENTED** [review]
Thanks! Looks great, added some small comments

File: test/test_nestedtensor.py (hunk 6 above)
Comment (soulitzer) [reply test/test_nestedtensor.py:7036]: Actually looking at CI, this is actually succeding :P
